src/ledger/LedgerManagerImpl.cpp

```diff
@@ -1573,6 +1576,414 @@ LedgerManagerImpl::prefetchTransactionData(AbstractLedgerTxnParent& ltx,
+                ttlExtension;
+            it->second.isDirty = true;
+        }
+    }
+
+    return threadRestoredKeys;
+}
+
+ParallelLedgerInfo
+getParallelLedgerInfo(AppConnector& app, AbstractLedgerTxn& ltx)
+{
+    auto const& lh = ltx.loadHeader().current();
+    return {lh.ledgerVersion, lh.ledgerSeq, lh.baseReserve,
+            lh.scpValue.closeTime, app.getNetworkID()};
+}
+
+void
+LedgerManagerImpl::applySorobanStage(AppConnector& app, AbstractLedgerTxn& ltx,
+                                     ApplyStage const& stage,
+                                     Hash const& sorobanBasePrngSeed)
+{
+    for (auto const& thread : stage)
+    {
+        for (auto const& txBundle : thread)
+        {
+            // This can mark a tx as failed due to validation.
+            // Make sure results are checked.
+            txBundle.getTx()->preParallelApply(app, ltx,
+                                               txBundle.getEffects().getMeta(),
+                                               txBundle.getResPayload());
+        }
+    }
+
+    auto const& config = app.getConfig();
+    auto const& sorobanConfig = app.getSorobanNetworkConfigForApply();
+    auto& sorobanMetrics = app.getLedgerManager().getSorobanMetrics();
+
+    std::vector<ThreadEntryMap> entryMapsByThread;
+    for (auto const& thread : stage)
+    {
+        entryMapsByThread.emplace_back(collectEntries(app, ltx, thread));
+    }
+
+    std::vector<std::future<RestoredKeys>> threadRestoredKeyFutures;
+
+    auto ledgerInfo = getParallelLedgerInfo(app, ltx);
+    std::vector<std::thread> threads;
+    for (size_t i = 0; i < stage.size(); ++i)
+    {
+        auto& entryMapByThread = entryMapsByThread.at(i);
+
+        auto const& thread = stage.at(i);
+
+        threadRestoredKeyFutures.emplace_back(std::async(
+            std::launch::async, &LedgerManagerImpl::applyThread, this,
+            std::ref(app), std::ref(entryMapByThread), std::ref(thread), config,
+            sorobanConfig, std::ref(ledgerInfo), sorobanBasePrngSeed,
+            std::ref(sorobanMetrics)));
+    }
+
+    for (auto& thread : threads)
+    {
+        thread.join();
+    }
+
+    std::vector<RestoredKeys> threadRestoredKeys;
+    for (auto& restoredKeysFutures : threadRestoredKeyFutures)
+    {
+        threadRestoredKeys.emplace_back(restoredKeysFutures.get());
+    }
+
+    LedgerTxn ltxInner(ltx);
+    for (auto const& restoredKeys : threadRestoredKeys)
+    {
+        // We filter out the TTL keys here because LedgerTxn will add the ttl
+        // key
+        for (auto const& key : restoredKeys.hotArchive)
+        {
+            if (key.type() != TTL)
+            {
+                ltxInner.addRestoredFromHotArchiveKey(key);
+            }
+        }
+        for (auto const& key : restoredKeys.liveBucketList)
+        {
+            if (key.type() != TTL)
+            {
+                ltxInner.addRestoredFromLiveBucketListKey(key);
```

Comment:
We should be able to get rid of `addRestoredFromLiveBucketListKey` and `LedgerTxn::mRestoredKeys.liveBucketList` entirely. The only reason we need to store liveBucketList restoration events in the ltx is for the meta post processing step. However, if we do that without the ltx, this is no longer necessary.

We still need to keep track of hotArchiveBucketList restores at the ltx level though. Maybe we can rename this to something like "removeKeyFromHotArchive." If we no longer generate restore meta via ltx, we just need to know what keys to delete from the HotArchive on ledger close.